### DISCUSSION TITLE
Eeprom loosing its settings occasionally

### DISCUSSION THREAD

**Discussion by Yveaux** [gcormier/megadesk#169] [Q&A] [answered]
On a self-assembled megadesk, which works just fine, I occasionally seem to loose the eeprom settings.
I regularly switch power supply of the desk.

Reading the Atmega's datasheet the most likely cause of eeprom corruption would be the execution of random code due to supply going down/coming up.

The BOD protection of the Atmega will keep the processor in reset when power levels are too low.
Looking at the fuses for megadesk, `fusee` is set to `0xFE`, which disables the BOD.
As the Attiny is powered by 5V, enabling BOD at 4.3V typically (`fusee` set to `0xF4`, `fuseh` set to `0xD4`) would be able to prevent this kind of random code execution.

I will perform some tests with BOD enabled, but wondered if there is a good reason it is disabled by default ?

--
Update: First tests show normal operation; now to see if settings are preserved over time
Update2: Used it for a week, and eeprom loss did no longer occur

**Comment by gcormier** [comment]
Interesting! Let me know how things go.

Do you have any references for the datasheet where it mentions not using BOD can cause eeprom corruption?

**Reply by Yveaux** [reply]
Yes, https://ww1.microchip.com/downloads/en/DeviceDoc/Atmel-8495-8-bit-AVR-Microcontrollers-ATtiny441-ATtiny841_Datasheet.pdf pg 19/20:
![image](https://github.com/gcormier/megadesk/assets/7716005/816141d5-5fbb-48e7-9738-ed1d5ba68c08)
![image](https://github.com/gcormier/megadesk/assets/7716005/e1b230f7-93fb-4fce-a8ef-0e65a7e0958a)
Especially the second case may be an issue for Megadesk, causing rogue code to execute that may write to eeprom.

**Comment by gcormier** [comment]
Thanks! I've pushed some updates based on your recommendation :) Do you plan on building a companion?

**Reply by Yveaux** [reply]
Nice, but I think you missed efuse: https://github.com/gcormier/megadesk/commit/543a6fb269b132be685ad595fe5b48fd1684b8ae#r144205644
See pg. 219 of the datasheet.

**Reply by Yveaux** [reply]
I didn't plan on building a companion, but I'd be happy to test it for you ;-)
Do you expect issues with it in combination with the BOD?

**Reply by gcormier** [reply]
No, but reach out to me on Tindie I'd love to support your efforts improving the project with the BOD observation.

**Comment by Yveaux** [comment, marked as answer]
After 2 weeks of testing I think it's safe to conclude that enabling BOD fixes the issue of eeprom loosing its settings.
Not a single loss occurred with BOD enabled, while before about every other day I had to re-store the desk levels.

**Comment by pfak** [comment]
This is fantastic news. This has been an issue for me for awhile. Is there a guide to updating? (I do not have a companion.)

**Reply by Yveaux** [reply]
AFAIK you need a programmer to do so.
This shows how to program the fuses for Megadesk: https://github.com/gcormier/megadesk/blob/master/DIY.md
And here you can find some more generic information on programming fuses: https://developerhelp.microchip.com/xwiki/bin/view/products/mcu-mpu/8-bit-avr/structure/fuses/
